File: ios/StillMoment/Domain/Models/MeditationSettings.swift
//
//  MeditationSettings.swift
//  Still Moment
//
//  Domain - Meditation Settings Model
//

import Foundation

/// Settings for meditation sessions
struct MeditationSettings: Codable, Equatable {
    // MARK: Lifecycle

    // MARK: - Initialization

    /// Default volume for background sounds (15%)
    static let defaultBackgroundSoundVolume: Float = 0.15

    /// Default volume for gong sounds (100%)
    static let defaultGongVolume: Float = 1.0

    /// Default volume for interval gong (75%)
    static let defaultIntervalGongVolume: Float = 0.75

    init(
        intervalGongsEnabled: Bool = false,
        intervalMinutes: Int = 5,
        intervalMode: IntervalMode = .repeating,
        intervalSoundId: String = GongSound.defaultIntervalSoundId,
        intervalGongVolume: Float = MeditationSettings.defaultIntervalGongVolume,
        backgroundSoundId: String = "silent",
        backgroundSoundVolume: Float = MeditationSettings.defaultBackgroundSoundVolume,
        durationMinutes: Int = 10,
        preparationTimeEnabled: Bool = true,
        preparationTimeSeconds: Int = 15,
        startGongSoundId: String = GongSound.defaultSoundId,
        gongVolume: Float = MeditationSettings.defaultGongVolume,
        attunementId: String? = nil,
        attunementEnabled: Bool = false,
        customAttunementDurationSeconds: Int? = nil
    ) {
        self.intervalGongsEnabled = intervalGongsEnabled
        self.intervalMinutes = Self.validateInterval(intervalMinutes)
        self.intervalMode = intervalMode
        self.intervalSoundId = intervalSoundId
        self.intervalGongVolume = Self.validateVolume(intervalGongVolume)
        self.backgroundSoundId = backgroundSoundId
        self.backgroundSoundVolume = Self.validateVolume(backgroundSoundVolume)
        self.customAttunementDurationSeconds = customAttunementDurationSeconds
        self.durationMinutes = Self.validateDuration(
            durationMinutes,
            attunementId: attunementId,
            attunementEnabled: attunementEnabled,
            attunementDurationSeconds: customAttunementDurationSeconds
        )
        self.preparationTimeEnabled = preparationTimeEnabled
        self.preparationTimeSeconds = Self.validatePreparationTime(preparationTimeSeconds)
        self.startGongSoundId = startGongSoundId
        self.gongVolume = Self.validateVolume(gongVolume)
        self.attunementId = attunementId
        self.attunementEnabled = attunementEnabled
    }

    // MARK: Internal

    // MARK: - Persistence Keys

    enum Keys {
        static let intervalGongsEnabled = "intervalGongsEnabled"
        static let intervalMinutes = "intervalMinutes"
        static let intervalMode = "intervalMode"
        static let intervalSoundId = "intervalSoundId"
        static let intervalGongVolume = "intervalGongVolume"
        static let backgroundSoundId = "backgroundSoundId"
        static let backgroundSoundVolume = "backgroundSoundVolume"
        static let durationMinutes = "durationMinutes"
        static let preparationTimeEnabled = "preparationTimeEnabled"
        static let preparationTimeSeconds = "preparationTimeSeconds"
        static let startGongSoundId = "startGongSoundId"
        static let gongVolume = "gongVolume"
        static let attunementId = "introductionId"
        static let attunementEnabled = "introductionEnabled"
        /// Legacy key for migration
        static let legacyBackgroundAudioMode = "backgroundAudioMode"
    }

    /// Whether interval gongs are enabled during meditation
    var intervalGongsEnabled: Bool

    /// Interval in minutes between gongs (1-60)
    var intervalMinutes: Int

    /// Interval gong mode (repeating, afterStart, beforeEnd)
    var intervalMode: IntervalMode

    /// Sound ID for interval gongs (independent from start/end gong)
    var intervalSoundId: String

    /// Interval gong volume (0.0 to 1.0)
    var intervalGongVolume: Float

    /// Background sound ID (references BackgroundSound.id)
    var backgroundSoundId: String

    /// Background sound volume (0.0 to 1.0)
    var backgroundSoundVolume: Float

    /// Duration of meditation in minutes (1-60)
    var durationMinutes: Int

    /// Whether preparation time is enabled before meditation starts
    var preparationTimeEnabled: Bool

    /// Duration of preparation phase in seconds (5, 10, 15, 20, 30, 45)
    var preparationTimeSeconds: Int

    /// Gong sound ID for start/end gong (references GongSound.id)
    var startGongSoundId: String

    /// Gong volume (0.0 to 1.0) - applies to start and end gong
    var gongVolume: Float

    /// Attunement ID (nil = no attunement, references Attunement.id)
    var attunementId: String?

    /// Whether the attunement is enabled (separate from attunementId to preserve user's selection)
    var attunementEnabled: Bool

    /// Custom attunement duration in seconds, resolved by ViewModel. Nil for built-in attunements.
    /// Transient — not persisted to UserDefaults.
    var customAttunementDurationSeconds: Int?

    /// The effective attunement ID. `nil` when disabled or no attunement is selected.
    /// Use this instead of checking `attunementEnabled` + `attunementId` manually.
    var activeAttunementId: String? {
        guard self.attunementEnabled else {
            return nil
        }
        return self.attunementId
    }

    // MARK: - Validation

    /// Validates and clamps interval to valid range (1-60 minutes)
    static func validateInterval(_ minutes: Int) -> Int {
        min(max(minutes, 1), 60)
    }

    /// Returns the minimum meditation duration in minutes for a given active attunement ID.
    /// `activeAttunementId` is `nil` when disabled or unset — callers use `settings.activeAttunementId`.
    /// `attunementDurationSeconds` is provided by the caller (resolved via AttunementResolver).
    /// Falls back to `Attunement.find()` for built-in attunements when `attunementDurationSeconds` is nil.
    /// Formula: ceil(attunementDurationSeconds / 60)
    static func minimumDuration(activeAttunementId: String?, attunementDurationSeconds: Int? = nil) -> Int {
        guard activeAttunementId != nil else {
            return 1
        }
        let durationSeconds: Int
        if let provided = attunementDurationSeconds {
            durationSeconds = provided
        } else if let attunementId = activeAttunementId,
                  let attunement = Attunement.find(byId: attunementId) {
            durationSeconds = attunement.durationSeconds
        } else {
            return 1
        }
        guard durationSeconds > 0 else {
            return 1
        }
        return Int(ceil(Double(durationSeconds) / 60.0))
    }

    /// Backward-compatible overload used during init/validation where enabled+id are separate.
    static func minimumDuration(
        for attunementId: String?,
        attunementEnabled: Bool = false,
        attunementDurationSeconds: Int? = nil
    ) -> Int {
        let activeId = attunementEnabled ? attunementId : nil
        return Self.minimumDuration(
            activeAttunementId: activeId,
            attunementDurationSeconds: attunementDurationSeconds
        )
    }

    /// Minimum meditation duration in minutes based on current attunement setting.
    /// Uses `customAttunementDurationSeconds` when set (resolved via AttunementResolver).
    var minimumDurationMinutes: Int {
        Self.minimumDuration(
            activeAttunementId: self.activeAttunementId,
            attunementDurationSeconds: self.customAttunementDurationSeconds
        )
    }

    /// Validates and clamps duration to valid range (minimum-60 minutes).
    /// Minimum is 1 without attunement, or ceil(attunementDuration/60) with enabled attunement.
    static func validateDuration(
        _ minutes: Int,
        attunementId: String? = nil,
        attunementEnabled: Bool = false,
        attunementDurationSeconds: Int? = nil
    ) -> Int {
        let minimum = Self.minimumDuration(
            for: attunementId,
            attunementEnabled: attunementEnabled,
            attunementDurationSeconds: attunementDurationSeconds
        )
        return min(max(minutes, minimum), 60)
    }

    /// Valid preparation time options in seconds
    static let validPreparationTimes = [5, 10, 15, 20, 30, 45]

    /// Validates and clamps preparation time to valid values (5, 10, 15, 20, 30, 45 seconds)
    static func validatePreparationTime(_ seconds: Int) -> Int {
        // Find the closest valid value
        self.validPreparationTimes.min { abs($0 - seconds) < abs($1 - seconds) } ?? 15
    }

    /// Validates and clamps volume to valid range (0.0-1.0)
    static func validateVolume(_ volume: Float) -> Float {
        min(max(volume, 0.0), 1.0)
    }
}

// MARK: - Default Settings

extension MeditationSettings {
    /// Default settings with interval gongs disabled and silent background audio
    static let `default` = MeditationSettings(
        intervalGongsEnabled: false,
        intervalMinutes: 5,
        intervalMode: .repeating,
        intervalSoundId: GongSound.defaultIntervalSoundId,
        intervalGongVolume: defaultIntervalGongVolume,
        backgroundSoundId: "silent",
        backgroundSoundVolume: defaultBackgroundSoundVolume,
        durationMinutes: 10,
        preparationTimeEnabled: true,
        preparationTimeSeconds: 15,
        startGongSoundId: GongSound.defaultSoundId,
        gongVolume: defaultGongVolume,
        attunementId: nil,
        attunementEnabled: false
    )
}

// MARK: - Legacy Migration

extension MeditationSettings {
    /// Migrates legacy BackgroundAudioMode enum to sound ID
    /// - Parameter mode: Legacy enum value
    /// - Returns: Corresponding sound ID
    static func migrateLegacyMode(_ mode: String) -> String {
        switch mode {
        case "Silent":
            "silent"
        case "White Noise":
            "silent" // WhiteNoise removed, fallback to silent
        default:
            "silent"
        }
    }
}

File: android/app/src/main/kotlin/com/stillmoment/domain/models/MeditationSettings.kt
package com.stillmoment.domain.models

/**
 * Settings for meditation sessions.
 *
 * @property intervalGongsEnabled Whether interval gongs are enabled during meditation
 * @property intervalMinutes Interval in minutes between gongs (1-60)
 * @property intervalMode How interval gongs are triggered (REPEATING, AFTER_START, BEFORE_END)
 * @property intervalSoundId ID of the sound for interval gongs (references GongSound.id, including "soft-interval")
 * @property intervalGongVolume Volume for interval gong sounds (0.0 to 1.0)
 * @property backgroundSoundId Background sound ID (references BackgroundSound.id)
 * @property durationMinutes Duration of meditation in minutes (1-60)
 * @property preparationTimeEnabled Whether preparation time before meditation is enabled
 * @property preparationTimeSeconds Duration of preparation in seconds (5, 10, 15, 20, 30, or 45)
 * @property gongSoundId ID of the gong sound for start/end (references GongSound.id)
 * @property gongVolume Volume for start/end gong sounds (0.0 to 1.0)
 * @property attunementEnabled Whether attunement audio is enabled
 * @property customAttunementDurationSeconds Duration of custom attunement in seconds (sync resolution path).
 *   Populated by TimerViewModel from CustomAudioRepository when converting Praxis to MeditationSettings.
 *   Built-in attunements use [Attunement.find] instead (no need for this field).
 *   The async resolution path (Resolver) provides the duration for UI display independently.
 */
data class MeditationSettings(
    val intervalGongsEnabled: Boolean = false,
    val intervalMinutes: Int = DEFAULT_INTERVAL_MINUTES,
    val intervalMode: IntervalMode = DEFAULT_INTERVAL_MODE,
    val intervalSoundId: String = DEFAULT_INTERVAL_SOUND_ID,
    val intervalGongVolume: Float = DEFAULT_INTERVAL_GONG_VOLUME,
    val backgroundSoundId: String = DEFAULT_BACKGROUND_SOUND_ID,
    val backgroundSoundVolume: Float = DEFAULT_BACKGROUND_SOUND_VOLUME,
    val durationMinutes: Int = DEFAULT_DURATION_MINUTES,
    val preparationTimeEnabled: Boolean = DEFAULT_PREPARATION_TIME_ENABLED,
    val preparationTimeSeconds: Int = DEFAULT_PREPARATION_TIME_SECONDS,
    val gongSoundId: String = DEFAULT_GONG_SOUND_ID,
    val gongVolume: Float = DEFAULT_GONG_VOLUME,
    val attunementId: String? = null,
    val attunementEnabled: Boolean = DEFAULT_ATTUNEMENT_ENABLED,
    val customAttunementDurationSeconds: Int? = null
) {
    init {
        // Validation is applied through copy() and create() methods
        // Direct construction allows any values for flexibility in deserialization
    }

    companion object {
        const val DEFAULT_INTERVAL_MINUTES = 5
        val DEFAULT_INTERVAL_MODE = IntervalMode.REPEATING
        const val DEFAULT_INTERVAL_SOUND_ID = GongSound.SOFT_INTERVAL_SOUND_ID
        const val DEFAULT_INTERVAL_GONG_VOLUME = 0.75f
        const val DEFAULT_BACKGROUND_SOUND_ID = "silent"
        const val DEFAULT_BACKGROUND_SOUND_VOLUME = 0.15f
        const val DEFAULT_DURATION_MINUTES = 10
        const val DEFAULT_PREPARATION_TIME_ENABLED = true
        const val DEFAULT_PREPARATION_TIME_SECONDS = 15
        const val DEFAULT_GONG_SOUND_ID = GongSound.DEFAULT_SOUND_ID
        const val DEFAULT_GONG_VOLUME = 1.0f
        const val DEFAULT_ATTUNEMENT_ENABLED = false

        /** Minimum interval in minutes */
        const val MIN_INTERVAL_MINUTES = 1

        /** Maximum interval in minutes */
        const val MAX_INTERVAL_MINUTES = 60

        // Valid preparation time options (in seconds)
        val VALID_PREPARATION_TIMES = listOf(5, 10, 15, 20, 30, 45)

        /** Default settings with interval gongs disabled and silent background audio */
        val Default = MeditationSettings()

        /**
         * Validates and clamps interval to valid range (1-60 minutes).
         */
        fun validateInterval(minutes: Int): Int {
            return minutes.coerceIn(MIN_INTERVAL_MINUTES, MAX_INTERVAL_MINUTES)
        }

        /**
         * Validates and clamps duration to valid range (1-60 minutes).
         * When an attunementId is provided and attunement is enabled,
         * enforces a minimum based on attunement duration.
         */
        fun validateDuration(
            minutes: Int,
            attunementId: String? = null,
            attunementEnabled: Boolean = false,
            customAttunementDurationSeconds: Int? = null,
        ): Int {
            val min = minimumDuration(attunementId, attunementEnabled, customAttunementDurationSeconds)
            return minutes.coerceIn(min, 60)
        }

        /**
         * Returns the minimum meditation duration in minutes for a given active attunement ID.
         * [activeAttunementId] is `null` when disabled or unset — callers use [MeditationSettings.activeAttunementId].
         * When [customAttunementDurationSeconds] is provided, it is used instead of looking up built-in attunements.
         * Formula: ceil(attunementDuration / 60)
         */
        fun minimumDuration(activeAttunementId: String?, customAttunementDurationSeconds: Int? = null): Int {
            if (activeAttunementId == null) return 1
            val durationSeconds = when {
                customAttunementDurationSeconds != null -> customAttunementDurationSeconds
                else -> Attunement.find(activeAttunementId)?.durationSeconds ?: return 1
            }
            if (durationSeconds <= 0) return 1
            return kotlin.math.ceil(durationSeconds / 60.0).toInt()
        }

        /**
         * Backward-compatible overload used during init/validation where enabled and id are separate.
         */
        fun minimumDuration(
            attunementId: String? = null,
            attunementEnabled: Boolean = false,
            customAttunementDurationSeconds: Int? = null,
        ): Int {
            val activeId = if (attunementEnabled) attunementId else null
            return minimumDuration(
                activeAttunementId = activeId,
                customAttunementDurationSeconds = customAttunementDurationSeconds,
            )
        }

        /**
         * Validates and clamps preparation time to nearest valid value (5, 10, 15, 20, 30, or 45).
         */
        fun validatePreparationTime(seconds: Int): Int {
            return VALID_PREPARATION_TIMES.minByOrNull { kotlin.math.abs(it - seconds) }
                ?: DEFAULT_PREPARATION_TIME_SECONDS
        }

        /**
         * Validates and clamps volume to valid range (0.0-1.0).
         */
        fun validateVolume(volume: Float): Float {
            return volume.coerceIn(0f, 1f)
        }

        /**
         * Migrates legacy BackgroundAudioMode enum to sound ID.
         *
         * @param mode Legacy enum value
         * @return Corresponding sound ID
         */
        fun migrateLegacyMode(mode: String): String {
            return when (mode) {
                "Silent" -> "silent"
                "White Noise" -> "silent" // WhiteNoise removed, fallback to silent
                else -> "silent"
            }
        }

        /**
         * Creates settings with validated values.
         */
        fun create(
            intervalGongsEnabled: Boolean = false,
            intervalMinutes: Int = DEFAULT_INTERVAL_MINUTES,
            intervalMode: IntervalMode = DEFAULT_INTERVAL_MODE,
            intervalSoundId: String = DEFAULT_INTERVAL_SOUND_ID,
            intervalGongVolume: Float = DEFAULT_INTERVAL_GONG_VOLUME,
            backgroundSoundId: String = DEFAULT_BACKGROUND_SOUND_ID,
            backgroundSoundVolume: Float = DEFAULT_BACKGROUND_SOUND_VOLUME,
            durationMinutes: Int = DEFAULT_DURATION_MINUTES,
            preparationTimeEnabled: Boolean = DEFAULT_PREPARATION_TIME_ENABLED,
            preparationTimeSeconds: Int = DEFAULT_PREPARATION_TIME_SECONDS,
            gongSoundId: String = DEFAULT_GONG_SOUND_ID,
            gongVolume: Float = DEFAULT_GONG_VOLUME,
            attunementId: String? = null,
            attunementEnabled: Boolean = DEFAULT_ATTUNEMENT_ENABLED,
            customAttunementDurationSeconds: Int? = null
        ): MeditationSettings {
            return MeditationSettings(
                intervalGongsEnabled = intervalGongsEnabled,
                intervalMinutes = validateInterval(intervalMinutes),
                intervalMode = intervalMode,
                intervalSoundId = intervalSoundId,
                intervalGongVolume = validateVolume(intervalGongVolume),
                backgroundSoundId = backgroundSoundId,
                backgroundSoundVolume = validateVolume(backgroundSoundVolume),
                durationMinutes = validateDuration(
                    durationMinutes,
                    attunementId,
                    attunementEnabled,
                    customAttunementDurationSeconds,
                ),
                preparationTimeEnabled = preparationTimeEnabled,
                preparationTimeSeconds = validatePreparationTime(preparationTimeSeconds),
                gongSoundId = gongSoundId,
                gongVolume = validateVolume(gongVolume),
                attunementId = attunementId,
                attunementEnabled = attunementEnabled,
                customAttunementDurationSeconds = customAttunementDurationSeconds
            )
        }
    }

    /**
     * Returns a copy with validated interval minutes.
     */
    fun withIntervalMinutes(minutes: Int): MeditationSettings {
        return copy(intervalMinutes = validateInterval(minutes))
    }

    /**
     * The effective attunement ID. `null` when disabled or no attunement is selected.
     * Use this instead of checking [attunementEnabled] + [attunementId] manually.
     */
    val activeAttunementId: String?
        get() = if (attunementEnabled) attunementId else null

    /**
     * Whether an active attunement is configured and available.
     * Returns true for custom attunements (when [customAttunementDurationSeconds] is set)
     * AND for built-in attunements (when available for current language).
     */
    val hasActiveAttunement: Boolean
        get() {
            if (!attunementEnabled) return false
            attunementId ?: return false
            // Custom attunement: customAttunementDurationSeconds is set
            if (customAttunementDurationSeconds != null) return true
            // Built-in: check language availability
            return Attunement.isAvailableForCurrentLanguage(attunementId)
        }

    /**
     * The effective attunement duration in seconds.
     * Returns [customAttunementDurationSeconds] for custom attunements, or the built-in duration
     * for standard attunements. Returns 0 if no active attunement.
     */
    val effectiveAttunementDurationSeconds: Int
        get() {
            if (!hasActiveAttunement) return 0
            // Custom attunement duration
            customAttunementDurationSeconds?.let { return it }
            // Built-in duration
            val id = attunementId ?: return 0
            return Attunement.find(id)?.durationSeconds ?: 0
        }

    /**
     * Returns the minimum duration in minutes based on the current attunement setting.
     * Uses [customAttunementDurationSeconds] when set (for custom attunements).
     */
    val minimumDurationMinutes: Int
        get() = minimumDuration(
            activeAttunementId = activeAttunementId,
            customAttunementDurationSeconds = customAttunementDurationSeconds
        )

    /**
     * Returns a copy with validated duration minutes (respects attunement minimum).
     * Uses [customAttunementDurationSeconds] when set (for custom attunements).
     */
    fun withDurationMinutes(minutes: Int): MeditationSettings {
        return copy(
            durationMinutes = validateDuration(
                minutes,
                attunementId,
                attunementEnabled,
                customAttunementDurationSeconds,
            )
        )
    }

    /**
     * Returns a copy with the attunement enabled or disabled.
     */
    fun withAttunementEnabled(enabled: Boolean): MeditationSettings {
        return copy(attunementEnabled = enabled)
    }

    /**
     * Returns a copy with validated preparation time seconds.
     */
    fun withPreparationTimeSeconds(seconds: Int): MeditationSettings {
        return copy(preparationTimeSeconds = validatePreparationTime(seconds))
    }
}

/**
 * Persistence keys for MeditationSettings.
 * Used with DataStore for storing settings.
 */
object MeditationSettingsKeys {
    const val INTERVAL_GONGS_ENABLED = "intervalGongsEnabled"
    const val INTERVAL_MINUTES = "intervalMinutes"
    const val INTERVAL_MODE = "intervalMode"
    const val INTERVAL_SOUND_ID = "intervalSoundId"
    const val INTERVAL_GONG_VOLUME = "intervalGongVolume"
    const val BACKGROUND_SOUND_ID = "backgroundSoundId"
    const val BACKGROUND_SOUND_VOLUME = "backgroundSoundVolume"
    const val DURATION_MINUTES = "durationMinutes"
    const val PREPARATION_TIME_ENABLED = "preparationTimeEnabled"
    const val PREPARATION_TIME_SECONDS = "preparationTimeSeconds"
    const val GONG_SOUND_ID = "gongSoundId"
    const val GONG_VOLUME = "gongVolume"
    const val ATTUNEMENT_ID = "introductionId"
    const val ATTUNEMENT_ENABLED = "introductionEnabled"

    // Legacy keys for migration
    const val LEGACY_BACKGROUND_AUDIO_MODE = "backgroundAudioMode"
    const val LEGACY_INTERVAL_REPEATING = "interval_repeating"
    const val LEGACY_INTERVAL_FROM_END = "interval_from_end"
}
